pkg/ottl/ottlfuncs/func_extract_grok_patterns.go

```diff
@@ -0,0 +1,98 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/elastic/go-grok"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type ExtractGrokPatternsArguments[K any] struct {
+	Target             ottl.StringGetter[K]
+	Pattern            string
+	NamedCapturesOnly  ottl.Optional[bool]
+	PatternDefinitions ottl.Optional[[]ottl.StringGetter[K]]
+}
+
+func NewExtractGrokPatternsFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("ExtractGrokPatterns", &ExtractGrokPatternsArguments[K]{}, createExtractGrokPatternsFunction[K])
+}
+
+func createExtractGrokPatternsFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ExtractGrokPatternsArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ExtractGrokPatternsFactory args must be of type *ExtractGrokPatternsArguments[K]")
+	}
+
+	return extractGrokPatterns(args.Target, args.Pattern, args.NamedCapturesOnly, args.PatternDefinitions)
+}
+
+func extractGrokPatterns[K any](target ottl.StringGetter[K], pattern string, nco ottl.Optional[bool], patternDefinitions ottl.Optional[[]ottl.StringGetter[K]]) (ottl.ExprFunc[K], error) {
+	g := grok.NewComplete()
+	namedCapturesOnly := !nco.IsEmpty() && nco.Get()
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		if !patternDefinitions.IsEmpty() {
+			getters := patternDefinitions.Get()
+
+			for i, getter := range getters {
+				val, err := getter.Get(ctx, tCtx)
+				if err != nil {
+					return nil, fmt.Errorf("failed to get pattern supplied to ExtractGrokPatterns at index %d: %w", i, err)
+				}
+				// split pattern in format key=val
+				parts := strings.SplitN(val, "=", 2)
+				if len(parts) == 1 {
+					return nil, fmt.Errorf("pattern supplied to ExtractGrokPatterns at index %d has incorrect format, expecting PATTERNNAME=pattern definition", i)
+				}
+
+				g.AddPattern(parts[0], parts[1])
```

Comment:
Do we need to check that `parts[0]` is formatted a specific way?

Reply:
no not really, convention is to use uppercase but it's not required, it should not contain `:` will add this check